### PR TITLE
Add bootloader unnamed module to root set

### PR DESCRIPTION
**Pull request by tajila**
Add bootloader unnamed module to root set

Also, modify stack walker to only look for module name if it is a named
module.

Related to: #14097

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@dmitripivkine please review GC changes

**Comment by tajila** [comment]
@pshipton please review stackwalker changes

**Comment by tajila** [comment]
This PR fixes the crash, but two of the tests will not pass due to our `-Dsun.reflect.noInflation=false` behaviour being different to the RI. I dont think this is a big deal as  `-Dsun.reflect.noInflation=true` is default, and our behvaiour has been different dating back to jdk11

**Comment by pshipton** [comment]
Pls refer to what is being fixed. What tests aren't passing? What is the proposal for dealing with the failures?

File: runtime/gc_vlhgc/WriteOnceCompactor.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2021 IBM Corp. and others
+ * Copyright (c) 1991, 2022 IBM Corp. and others
```
Comment (pshipton) [inline runtime/gc_vlhgc/WriteOnceCompactor.cpp:2]: The only change here is the copyright. This change should be removed.

**Comment by tajila** [comment]
> Pls refer to what issue is being fixed. What tests aren't passing? What is the proposal for dealing with the failures?

@pshipton Sorry, I had it in the commit but not the PR description. I've updated it.

For the remaining test failures we can exclude them. Moving forward `-Djdk.reflect.useDirectMethodHandle=false` will be phased out so I dont think its worth matching RI behaviour that will be deprecated.

**Comment by dmitripivkine** [comment]
This is kind of hack, better solution would be discover unnamed module in `classLoader->moduleHashTable`. I guess there is difficulty because it "unnamed". Is it make sense to look to this solution again - like use something else (artificial unique identifier ?) as a hash key?

In case the preference is to handle unnamed module explicitly why you does not add it finally to 
`MM_WriteOnceCompactor::fixupClassLoaderObject()` too?

**Comment by tajila** [comment]
> This is kind of hack, better solution would be discover unnamed module in classLoader->moduleHashTable. I guess there is difficulty because it "unnamed". Is it make sense to look to this solution again - like use something else (artificial unique identifier ?) as a hash key?

The unnamed loader cannot be put in the hash table because the module hash table uses a name as a key, but the unnamed loader doesnt have one.

**Comment by dmitripivkine** [comment]
> The unnamed loader cannot be put in the hash table because the module hash table uses a name as a key, but the unnamed loader doesnt have one.

Yes, I understand. It is why I mention using different (artificial) key for example. Well, if it is hard to do I am ok with this explicit solution.

Is `module->version` slot is not set too?

**Comment by tajila** [comment]
> Is module->version slot is not set too?

No name and version are always NULL

**Comment by pshipton** [comment]
> For the remaining test failures we can exclude them.

Is there a PR created for this yet?

**Comment by tajila** [comment]
> Is there a PR created for this yet?

No, I believe all the tests are excluded currently. 

We will need to enable 

```
jdk/internal/reflect/Reflection/GetCallerClassTest.java#id0 https://github.com/eclipse-openj9/openj9/issues/13998 generic-all
jdk/internal/reflect/Reflection/GetCallerClassTest.java#id1 https://github.com/eclipse-openj9/openj9/issues/13998 generic-all
```

Id2 & 3 test the `-Djdk.reflect.useDirectMethodHandle=false` behaviour

File: runtime/gc_glue_java/ConcurrentMarkingDelegate.cpp [hunk 2]
```diff
@@ -431,6 +431,10 @@ MM_ConcurrentMarkingDelegate::concurrentClassMark(MM_EnvironmentBase *env, bool
 					}
 				}
 
+				if (classLoader == _javaVM->systemClassLoader) {
+					_markingScheme->markObject(env, (j9object_t)_javaVM->unamedModuleForSystemLoader->moduleObject);
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/ConcurrentMarkingDelegate.cpp:435]: Do we need explicit cast to `(j9object_t)`? I think `J9Module-> moduleObject` has this type already

File: runtime/gc_vlhgc/GlobalMarkingScheme.cpp [hunk 3]
```diff
@@ -845,6 +845,12 @@ MM_GlobalMarkingScheme::scanClassLoaderObject(MM_EnvironmentVLHGC *env, J9Object
 				modulePtr = (J9Module**)hashTableNextDo(&walkState);
 			}
 		}
+
+		J9JavaVM *javaVM = ((J9VMThread*)env->getLanguageVMThread())->javaVM;
```
Comment (dmitripivkine) [inline runtime/gc_vlhgc/GlobalMarkingScheme.cpp:849]: _javaVM is defined in this class already, please use it instead

File: runtime/gc_vlhgc/WriteOnceCompactor.cpp [hunk 4]
```diff
@@ -1266,6 +1266,7 @@ MM_WriteOnceCompactor::fixupClassLoaderObject(MM_EnvironmentVLHGC* env, J9Object
 		 */
 		if (NULL != classLoader->moduleHashTable) {
 			J9HashTableState walkState;
+			J9JavaVM *javaVM = ((J9VMThread*)env->getLanguageVMThread())->javaVM;
```
Comment (dmitripivkine) [inline runtime/gc_vlhgc/WriteOnceCompactor.cpp:1269]: _javaVM is defined in this class already, please use it instead

**Review by dmitripivkine — APPROVED** [review]
Looks good to me with minor cosmetic changes

**Comment by tajila** [comment]
Changes ready for another look

**Comment by tajila** [comment]
@pshipton Please review the changes

**Comment by pshipton** [comment]
jenkins test sanity,sanity.openjdk alinux64 jdk11,jdk17

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
jenkins test sanity win jdk8

**Comment by tajila** [comment]
> Is this going to gpf on jdk8 trying to deference unamedModuleForSystemLoader ?

Yes, I'll need to fix that

**Comment by pshipton** [comment]
jdk 11, 17 testing https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/1478/ - passed

jenkins compile win jdk11

**Comment by pshipton** [comment]
jenkins test sanity alinux64 jdk8

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Review by pshipton — APPROVED** [review]

**Pull request merged**